object/slicer/slicer.go

```diff
@@ -0,0 +1,596 @@
+// specified limit, Slicer applies the slicing algorithm described within the
+// same specification. The outcome will be a group of "small" objects containing
+// a chunk of data, as well as an auxiliary linking object. All derived objects
+// are written to the parameterized ObjectWriter. If the amount of data is
+// within the limit, one object is produced. Note that Slicer can write multiple
+// objects, but returns the root object ID only.
+//
+// If current NeoFS epoch is specified via Options.SetCurrentNeoFSEpoch, it is
+// written to the metadata of all resulting objects as a creation epoch.
+//
+// See also NewSession.
+func New(signer neofscrypto.Signer, cnr cid.ID, owner user.ID, w ObjectWriter, opts Options) *Slicer {
+	return &Slicer{
+		signer: signer,
+		cnr:    cnr,
+		owner:  owner,
+		w:      w,
+		opts:   opts,
+	}
+}
+
+// NewSession creates Slicer which generates objects within provided session.
+// NewSession work similar to New with the detail that the session issuer owns
+// the produced objects. Specified session token is written to the metadata of
+// all resulting objects. In this case, the object is considered to be created
+// by a proxy on behalf of the session issuer.
+func NewSession(signer neofscrypto.Signer, cnr cid.ID, token session.Object, w ObjectWriter, opts Options) *Slicer {
+	return &Slicer{
+		signer:       signer,
+		cnr:          cnr,
+		owner:        token.Issuer(),
+		w:            w,
+		opts:         opts,
+		sessionToken: &token,
+	}
+}
+
+// childPayloadSizeLimit returns configured size limit of the child object's
+// payload which defaults to 1MB.
+func (x *Slicer) childPayloadSizeLimit() uint64 {
+	if x.opts.objectPayloadLimit > 0 {
+		return x.opts.objectPayloadLimit
+	}
+	return 1 << 20
+}
+
+// Slice creates new NeoFS object from the input data stream, associates the
+// object with the configured container and writes the object via underlying
+// ObjectWriter. After a successful write, Slice returns an oid.ID which is a
+// unique reference to the object in the container. Slice sets all required
+// calculated fields like payload length, checksum, etc.
+//
+// Slice allows you to specify string key-value pairs to be written to the
+// resulting object's metadata as object attributes. Corresponding argument MUST
+// NOT be empty or have odd length. Keys SHOULD NOT start with system-reserved
+// '__NEOFS__' prefix.
+//
+// See New for details.
+func (x *Slicer) Slice(data io.Reader, attributes ...string) (oid.ID, error) {
+	if len(attributes)%2 != 0 {
+		panic("attributes must be even number of strings")
+	}
+
+	var rootID oid.ID
+
+	fillCommonMetadata := func(obj *object.Object) {
+		currentVersion := version.Current()
+		obj.SetVersion(&currentVersion)
+		obj.SetContainerID(x.cnr)
+		obj.SetCreationEpoch(x.opts.currentNeoFSEpoch)
+		obj.SetType(object.TypeRegular)
+		obj.SetOwnerID(&x.owner)
+		obj.SetSessionToken(x.sessionToken)
+	}
```

Comment:
Good candidate to make this a part of the Slicer struct func, non anonymous one

Reply:
i don't mind